### PR TITLE
test: add tests to ensure compactor plans to not do unnecessary sorts

### PR DESCRIPTION
**Pull request by alamb**
# Rationale:

re https://github.com/influxdata/idpe/issues/16500

It turns out we rely a great deal on the compactor not resorting data to avoid OOMs in production, but we did not have coverage for that and a recent upgrade to datafusion lead to https://github.com/influxdata/idpe/issues/16500

# Changes

https://github.com/influxdata/influxdb_iox/pull/6196/files?w=1 I think is easier to review

Add test coverage based on the great comments of @NGA-TRAN in https://github.com/influxdata/influxdb_iox/pull/6160/files#r1025819594

However, I had to rearrange the code somewhat in order to be able to properly test the compactor plan (and assert there were no Sorts) and rather than hack in an extra level I refactored it into a builder style API. 

As part of this I removed a substantial amount of test setup duplication which I think makes it easier to see what is going on / being tested

"I hope you like what I have done with the place"

# Testing
I tested this by cherry-picking the commit to https://github.com/influxdata/influxdb_iox/pull/6160 and rerunning the tests. They fail like the following (as expected):

```shell
alamb@MacBook-Pro-8 influxdb_iox % git cherry-pick 196455ee9b9f3f9d5e28b7fb88e9a4d346a03e4a
git cherry-pick 196455ee9b9f3f9d5e28b7fb88e9a4d346a03e4a
[alamb/revert_revert_df_upgrade 90c1d7f83] refactor: Refactor ParquetFileCombining into a builder and plan, and add sort exec test

cargo test -p compactor -- parquet_file_combining
...
running 9 tests
test parquet_file_combining::tests::test_cutoff_bytes ... ok
test parquet_file_combining::tests::many_sorted_files_do_not_lose_data ... ok
test parquet_file_combining::tests::no_input_files_is_an_error ... ok
test parquet_file_combining::tests::one_input_file_gets_compacted ... ok
test parquet_file_combining::tests::compact_final_no_splits_creates_one_level_2_file ... FAILED
test parquet_file_combining::tests::small_files_get_compacted_into_one ... FAILED
test parquet_file_combining::tests::medium_input_files_cant_split_dont_make_empty_file ... FAILED
test parquet_file_combining::tests::medium_input_files_get_split_into_two ... FAILED
test parquet_file_combining::tests::large_input_files_get_split_multiple_times ... FAILED

failures:

---- parquet_file_combining::tests::compact_final_no_splits_creates_one_level_2_file stdout ----
Checking plan: SortPreservingMergeExec: [tag1@1 ASC,tag2@2 ASC,tag3@3 ASC,time@4 ASC]
  DeduplicateExec: [tag1@1 ASC,tag2@2 ASC,tag3@3 ASC,time@4 ASC]
    SortPreservingMergeExec: [tag1@1 ASC,tag2@2 ASC,tag3@3 ASC,time@4 ASC]
      SortExec: [tag1@1 ASC,tag2@2 ASC,tag3@3 ASC,time@4 ASC]
        UnionExec
          ParquetExec: limit=None, partitions=[1/1/1/1/10f799db-96f4-495e-8dde-0882483b60f0.parquet], output_ordering=[tag1@1 ASC, time@4 ASC], projection=[field_int, tag1, tag2, tag3, time]
          ParquetExec: limit=None, partitions=[1/1/1/1/eee2eee0-1678-4bd0-b8cc-955e714f7745.parquet], output_ordering=[tag2@2 ASC, tag3@3 ASC, time@4 ASC], projection=[field_int, tag1, tag2, tag3, time]

thread 'parquet_file_combining::tests::compact_final_no_splits_creates_one_level_2_file' panicked at 'Expected no sorts in compaction plan!', compactor/src/parquet_file_combining.rs:1791:13
stack backtrace:
   0: rust_begin_unwind
```

### CODE REVIEW

File: compactor/src/parquet_file_combining.rs [hunk 1]
```diff
@@ -704,285 +913,348 @@ mod tests {
...
+                    Arc::clone(&self.catalog.object_store),
+                    StorageId::from("iox"),
+                ))
+                .with_time_provider(Arc::clone(&self.catalog.time_provider) as _)
+                .with_compaction_input_file_bytes(self.metric.clone())
+                .with_max_desired_file_size_bytes(DEFAULT_MAX_DESIRED_FILE_SIZE_BYTES)
+                .with_percentage_max_file_size(DEFAULT_PERCENTAGE_MAX_FILE_SIZE)
+                .with_split_percentage(DEFAULT_SPLIT_PERCENTAGE)
+                // Verify that no compactor plans have Sorts!
+                .with_plan_observer(Box::new(assert_no_sorts_in_plan))
```
Comment (alamb) [inline compactor/src/parquet_file_combining.rs:935]: The whole point of this PR is to be able to add this observer that verifies there are no `SortExec`s in the physical plan

File: compactor/src/parquet_file_combining.rs [hunk 2]
```diff
@@ -704,285 +913,348 @@ mod tests {
...
+        sample_count: u64,
+        buckets_with_counts: Vec<(u64, u64)>,
     }
 
     #[tokio::test]
     // Cover multiple sorted files and that merging does not lose any data accidenally
     async fn no_input_files_is_an_error() {
         test_helpers::maybe_start_logging();
 
-        let TestSetup {
```
Comment (alamb) [inline compactor/src/parquet_file_combining.rs:958]: This test shows the pattern I applied to all the other tests: pulled much of the replicated code into `TestSetup` which I think makes it clearer what the test is actually testing -- none of the checks are changed, only the setup code

File: compactor/src/lib.rs [hunk 3]
```diff
@@ -388,38 +396,32 @@ pub(crate) async fn compact_one_partition(
             .await
             .context(UpgradingSnafu)?;
     } else if split {
-        parquet_file_combining::compact_parquet_files(
```
Comment (alamb) [inline compactor/src/lib.rs:391]: This file demonstrates the external API changes to a builder style API -- the biggest change is that the builder methods  I think document the parameter names a bit better.

File: compactor/src/parquet_file_combining.rs [hunk 4]
```diff
@@ -68,167 +69,290 @@ pub(crate) enum Error {
...
+        let num_files = files.len();
+        ensure!(
+            num_files > 0,
+            NotEnoughParquetFilesSnafu {
+                num_files,
+                partition_id
+            }
+        );
+
+        // check parameters
```
Comment (alamb) [inline compactor/src/parquet_file_combining.rs:225]: these checks are now necessary due to the builder API - it is unfortunate but I think straight forward

File: compactor/src/parquet_file_combining.rs [hunk 5]
```diff
@@ -68,167 +69,290 @@ pub(crate) enum Error {
...
+    #[snafu(display("Internal error: CompactPlanBuilder for partition {} missing required parameter: {}",
+                    partition_id.get(), name))]
+    BuilderParameter {
+        partition_id: PartitionId,
+        name: String,
+    },
 }
 
-// Compact the given parquet files received from `filter_parquet_files` into one stream
-#[allow(clippy::too_many_arguments)]
```
Comment (alamb) [inline compactor/src/parquet_file_combining.rs:74]: bonus is that clippy no longer needs to be ignored 👍

**Comment by NGA-TRAN** [comment]
@carols10cents and @lukebond : I am reviewing this but can you help review it, too? A lot of changes on the code that Carl knows well and Luke is changing them too

File: compactor/src/parquet_file_combining.rs [hunk 6]
```diff
@@ -68,167 +69,290 @@ pub(crate) enum Error {
...
+            name: "percentage_max_file_size",
+        })?;
+        let split_percentage = split_percentage.context(BuilderParameterSnafu {
+            partition_id,
+            name: "split_percentage",
+        })?;
+        let target_level = target_level.context(BuilderParameterSnafu {
+            partition_id,
+            name: "target_level",
+        })?;
```
Comment (NGA-TRAN) [inline compactor/src/parquet_file_combining.rs:263]: Yeah, this is the trade off I mentioned above. It is better to check them here

**Review by NGA-TRAN — APPROVED** [review]
The rewrite makes sense to me. Since a lot of code is being moved around, I have to rely on tests to make sure things are working as expected. We do have a lot of tests that I believe cover most (if not all) what we need. So I am confident this code won't break anything critical if any.

I think we may want to have this merged in and wait for a day or so to observe if the behaviors change on eu-central before we merge other Compactor related PRs. CC @lukebond

File: compactor/src/lib.rs [hunk 7]
```diff
@@ -388,38 +396,32 @@ pub(crate) async fn compact_one_partition(
...
+            .with_exec(Arc::clone(&compactor.exec))
+            .with_time_provider(Arc::clone(&compactor.time_provider))
+            .with_compaction_input_file_bytes(compactor.compaction_input_file_bytes.clone())
+            .with_max_desired_file_size_bytes(compactor.config.max_desired_file_size_bytes)
+            .with_percentage_max_file_size(compactor.config.percentage_max_file_size)
+            .with_split_percentage(compactor.config.split_percentage)
+            .with_target_level(target_level)
+            .build_with_splits()?
+            .compact_and_update_catalog()
+            .await?;
```
[the remark this answers is not in the document]
Comment (alamb) [reply compactor/src/lib.rs:412]: yes, that is right

File: compactor/src/parquet_file_combining.rs (hunk 6 above)
Comment (alamb) [reply compactor/src/parquet_file_combining.rs:263]: Yes, you are correct -- this is the tradeoff -- it was in some ways nicer to have the compiler help check that all parameters were specified. But I can't think of any other way around it really

**Comment by alamb** [comment]
I plan to wait until tomorrow for feedback from other reviewers prior to merging this in

**Pull request merged**